### PR TITLE
BST Complete

### PR DESCRIPTION
**Pull request by bv-yash**
Please review the code. @Ankit

### CODE REVIEW

File: bst.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+3,1,4,0,2,
```
Comment (bv-ankit) [inline bst.txt:1]: "," at the end not preferred.

Comment (bv-ankit) [reply bst.txt:1]: Blogvault learning.txt not needed in this repo

Comment (bv-ankit) [reply bst.txt:1]: Add a readme.md, with basic info about the project and how to use it

File: BST.rb [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+        return  @root.nil?
+    end
+
+end
+
+
+
+def show_oprations
+    puts "
+    1  => Add elements into the tree(multiple elements comma separated)
```
Comment (bv-ankit) [inline BST.rb]: This should be stored in a module.

File: BST.rb [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+      @value = value
+      @left = left
+      @right = right
+    end
+end
+
+
+class BinarySearchTree
+    
+    def initialize()
```
Comment (bv-ankit) [inline BST.rb]: "()" Not needed when not passing any params

File: BST.rb [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        @root = nil
```
Comment (bv-ankit) [inline BST.rb:18]: Add root as attr_accessor, then no need to do "@root", you can simply use "root".

File: BST.rb (hunk 2 above)
Comment (bv-ankit) [reply BST.rb]:
```
OPERATIONS = {}
def add_operation(id, msg)
    OPERATIONS[id] = msg
    id
end

module Operation
    SMALLEST_ELEMENT = add_operation(1, "Smallest Element: ")
    LARGEST_ELEMENT 
    ...
    ...
end
```

File: BST.rb [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+
+        elements.each do |elem|
+            tree.insert elem
+        end 
+        puts "Elements inserted successfully."
+
+    when "2"
+        puts "Largest Element : #{tree.find_largest}"
+
+    when "3"
```
Comment (bv-ankit) [inline BST.rb]: `when Operation::SMALLEST_ELEMENT`

File: BST.rb [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        level_order(@root)
+        puts
+    end
+
+
+    def in_order(root)
+        return if root.nil?
+            
+        in_order(root.left)
+
```
Comment (bv-ankit) [inline BST.rb]: Spacing not needed here, rethink spacing in the whole file.
Space should be used to separate logically different parts.

File: BST.rb [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+            tree.save
+            break
+        end
+            
+        run_script(tree, user_input)
+
+        puts "Press enter to continue." 
+        gets
+        
+        system('clear')
```
Comment (bv-ankit) [inline BST.rb]: In general, avoid using the system and exec commands.
It is fine for now.

File: BST.rb [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+    6  => Remove an element
+    7  => Print all the paths i.e starting from root to the leaf
+    8  => load BST from file.
+    9  => Save Tree
+    10 => Quit
+    "
+end
+
+def run_script(tree, user_input)
+    return puts "Tree is empty" if tree.empty? and user_input != "1"  and user_input != "8" and user_input > 10
```
Comment (bv-ankit) [inline BST.rb]: Handle this in `when case` itself.

File: BST.rb [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+def main
+
+    tree = BinarySearchTree.new()
+
+    while true
+
+        show_oprations
+
+        user_input = gets.chomp
+        if user_input == 'q' or user_input == 'quit' or user_input == '10'
```
Comment (bv-ankit) [inline BST.rb]: Handle this in the `run_script` itself.

File: BST.rb [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+        tree.load
+
+    when "9"
+        tree.save
+
+    else
+        puts 'Invalid Input !!!'
+    end
+end
+
```
Comment (bv-ankit) [inline BST.rb]: Mind spacing

File: main.rb [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    if user_input == 'quit'
+        user_input = 10
+    else
+        user_input = user_input.to_i
+    end
+
+    case user_input
+    when BstOperation::INSERT_ELEMENT
+        print "Please enter elemets : "
+        elements = gets.chomp.split(',').map { |elem| elem.to_i }
```
Comment (bv-ankit) [inline main.rb]:
```
def is_valid_input(input)
    input.to_i.to_s == input
end
```

File: main.rb [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    when BstOperation::LOAD
+        print 'Please enter file name : '
+        filename = gets.chomp
+        tree.load filename
+    when BstOperation::SAVE
+        return puts 'Tree is Empty' if tree.empty?
+
+        print 'Please enter file name : '
+        filename = gets.chomp
+        tree.save(filename)
```
Comment (bv-ankit) [inline main.rb:79]: store in json

File: main.rb [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        puts "Press enter to continue." 
+        gets
+        system('clear')
+
+    break if user_input == 'quit' or user_input.to_i == BstOperation::QUIT
+    end
+
+end
+
+
```
Comment (bv-ankit) [inline main.rb]: 2 line breaks not needed

File: main.rb [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        print op, ' => ', BST_OPERATION[op]
+        puts
+    end
+    print 'Please enter your choice : '
+end
+
+def run_bst_script(tree, user_input)
+
+    if user_input == 'quit'
+        user_input = 10
```
Comment (bv-ankit) [inline main.rb]: `user_input = (user_input == 'quit') ? 10 : user_input.to_i`

File: BST.rb [hunk 15]
```diff
@@ -0,0 +1,267 @@
+load 'Node.rb'
+
+class TreeNode < Node
+
+    attr_accessor :left, :right
```
Comment (bv-ankit) [inline BST.rb:5]: root can also be added here

File: BST.rb [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+        @left = left
+        @right = right
+    end
+end
+
+
+class BinarySearchTree
+    
+    def initialize
+        @root = nil
```
Comment (bv-ankit) [inline BST.rb:20]: this can be an attr_accessor

Comment (bv-ankit) [reply BST.rb:20]: not necessary for now

File: BST.rb [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+    end
+
+    def insert(value)
+        if @root == nil
```
Comment (bv-ankit) [inline BST.rb]: .nil?
nor necessary for now

File: main.rb [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    ALL_PATH = add_operation(7, 'All Paths i.e Root -> leaf')
+    LOAD = add_operation(8, 'Load Tree')
+    SAVE = add_operation(9, 'Save Tree')
+    QUIT = add_operation(10, 'Quit !')
+end
+    
+def show_bst_operations
+    total_op = BST_OPERATION.length
+
+    (1..total_op).each do |op|
```
Comment (bv-ankit) [inline main.rb]: BST_OPERATION.each do |key, value|
end

File: main.rb [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+end
+
+def run_bst_script(tree, user_input)
+
+    if user_input == 'quit'
+        user_input = 10
+    else
+        user_input = user_input.to_i
+    end
+
```
Comment (bv-ankit) [inline main.rb:44]: ops = [BstOperation::INSERT, BstOperation::SAVE, BstOperation::QUIT]
puts "Empty Tree" if tree.empty? && ops.include?(input)

File: main.rb [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+def main_BST
+    tree = BinarySearchTree.new()
+    
+    loop do
+        show_bst_operations
+        user_input = gets.chomp
+
+        run_bst_script(tree, user_input)
+
+        puts "Press enter to continue."
```
Comment (bv-ankit) [inline main.rb]:
```
def wait_for_user_input
    puts "Press enter to continue." 
    gets
    system('clear')
end
```

File: main.rb [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+        run_bst_script(tree, user_input)
+
+        puts "Press enter to continue." 
+        gets
+        system('clear')
+
+    break if user_input == 'quit' or user_input.to_i == BstOperation::QUIT
+    end
+
```
Comment (bv-ankit) [inline main.rb]: empty line not needed

File: main.rb (hunk 14 above)
Comment (bv-ankit) [reply main.rb]: this should be outside this function

File: main.rb [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+        show_bst_operations
+        user_input = gets.chomp
+
+        run_bst_script(tree, user_input)
+
+        puts "Press enter to continue." 
+        gets
+        system('clear')
+
+    break if user_input == 'quit' or user_input.to_i == BstOperation::QUIT
```
Comment (bv-ankit) [inline main.rb]: `break if user_input.to_i == BstOperation::QUIT`

File: BST.rb [hunk 23]
```diff
@@ -0,0 +1,267 @@
...
+        end
+
+        false
+    end
+
+    def remove(key)
+        @root = remove_helper(key, @root)
+    end
+
+    def remove_helper(key, node)
```
Comment (bv-ankit) [inline BST.rb]: `def remove_helper(node, key)`

File: BST.rb [hunk 24]
```diff
@@ -0,0 +1,267 @@
+load 'Node.rb'
+
```
Comment (bv-ankit) [inline BST.rb:3]: Once do a self review according to the coding guideleines

File: main.rb [hunk 25]
```diff
@@ -21,75 +21,82 @@ module BstOperation
...
-        user_input = 10
-    else
-        user_input = user_input.to_i
+def is_valid_input(elements)
+    begin
+        elements = elements.map { |elem| Integer(elem) }
+    rescue ArgumentError
+        return false
     end
+    true
```
Comment (bv-ankit) [inline main.rb]: this should be before rescue line

File: main.rb [hunk 26]
```diff
@@ -21,75 +21,82 @@ module BstOperation
...
+end
+
+def run_bst_script(tree, user_input)
+    
+    ops = [BstOperation::INSERT, BstOperation::LOAD,BstOperation::SAVE, BstOperation::QUIT]
+    return puts "Empty Tree" if tree.empty? and !ops.include?(user_input)
```
Comment (bv-ankit) [inline main.rb:43]:
```
if tree.empty? and !ops.include?(user_input)
    puts "Empty Tree"    
    return 
end
```

File: main.rb [hunk 27]
```diff
@@ -21,75 +21,82 @@ module BstOperation
...
-        return puts 'Tree is Empty' if tree.empty?
-
         print 'Please enter file name : '
         filename = gets.chomp
         tree.save(filename)
     when BstOperation::QUIT
         return puts 'Thank You!!!' if tree.empty?
 
-        print 'Please enter file name to save : '
+        print 'Please enter file name to save : '
```
Comment (bv-ankit) [inline main.rb]: fix indentation at the end of line

File: BST.rb [hunk 28]
```diff
@@ -13,19 +14,21 @@ def initialize(value = nil, left = nil, right = nil)
 
 
 class BinarySearchTree
+
+    attr_accessor :root
```
Comment (bv-ankit) [inline BST.rb:17]: If you have added this, change all the `@root` to `root` and any `root` inside a function can be `_root` or `curr_root`.

Comment (bv-ankit) [reply BST.rb:17]: https://www.rubyguides.com/2018/11/attr_accessor/

File: main.rb [hunk 29]
```diff
@@ -1,4 +1,4 @@
-load "BST.rb"
+require_relative "BST.rb"
```
Comment (bv-ankit) [inline main.rb]: In another file you have omitted `.rb`, keep it consistent according to the best practice.

File: main.rb (hunk 26 above)
Comment (bv-yash) [reply main.rb:43]: is there any problem with the single liner here?

File: BST.rb (hunk 28 above)
Comment (bv-yash) [reply BST.rb:17]: after adding attr_accessor, on changing @root -> root, breaks the working of code. need to discuss that with you.

**Comment by bv-ankit** [comment]
We use tab=8 spaces
Reformat the file

**Pull request merged**